include correct headers

LinusU opened the pull request:
This fixes so that each files includes the headers that it needs. This usually works anyway on `gcc` but `clang` is more stricter about this.

It also changes `_stricmp` to `strcasecmp` which I believe to be right based on this: http://stackoverflow.com/questions/1784767/g-error-stricmp-was-not-declared-in-this-scope-but-ok-for-strcmp

IntelOrca commented:
Why do we need to include limits in all these C files, where it can just be in common.h?

LinusU commented:
@janisozaur I'm still a beginner at all this so sorry if I made something in a bad way. What do you mean with "fix the platform defines as well"?

I realize now that the headers are indeed the c-headers instead of c++ one, I'll fix that.

Re. `strcasecmp` vs `strncasecmp`, since it used to be `_stricmp` I chose the functionally equivalent. It's to my understanding that, and I might be totally wrong, `_stricmp` isn't available because it isn't part of POSIX nor ANSI and thus wasn't available on my target (OS X + clang). But `strcasecmp` isn't available on windows so that won't work anyhow. [Relevant StackOverflow](http://stackoverflow.com/questions/11635/case-insensitive-string-comparison-in-c)

Re. including in the specific files instead of including it all in `common.h`. It's to my understanding that best practice is to include the necessary headers in each file. Here is a thread on StackOverflow which discusses this, especially note the most downvoted answer which suggests putting everything in `common.h`. It seems like the reasoning is that since you always include everything you end up wasting a lot of resources. [Link](http://stackoverflow.com/questions/25870961/c-do-i-have-to-include-standard-libraries-for-every-source-file)

I'm trying to get this compiling on OS X and this is the current issue I'm running into, would love to contribute back so that mac builds starts working. Happy to learn best practices as well!

IntelOrca commented:
Limits is a fundamental header file and will be needed by most files anyway which justifies it being in common.h. It is also possible to increase compilation times by making common.h a pre-compiled header.

janisozaur commented:
@LinusU have a look here: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/platform/platform.h#L162
That's a hack (the "platform define" I mentioned earlier) I made to circumvent functions missing. It's working, but not quite really the way you would do for a "production" code.
For the moment, you can try doing similar thing for OS X, this could possibly boost you past undefined functions.

@IntelOrca @LinusU personally I'm against globally including headers and would much rather limit them as much as possible. A pass of that in on my mental to-do list. This is partially based on my experience where a project's compilation time was brought down to 10% of original time after cleaning up the headers.

LinusU commented:
@janisozaur Just to clarify, you are against putting everything in `common.h` and would rather have it in the file where it's used? I think that's the best approach as well.

janisozaur commented:
With the latest commit I'm very curious: does it work? Can you `mmap` properly?

You would probably want to enable Linux platform file as well for OS X, at least for development, I suppose. Be cautious, however, it is not yet complete.

You may want to have a look at `install.sh`, `build.sh` and `.travis.yml` scripts. Especially regarding that part, could you provide clang build entry for travis?

LinusU commented:
With the addition of my second commit now it actually compiles with clang on OS X. There is some problem with the linking thought, and I have no idea on how to solve it. Any ideas would be appreciated :100: 

```text
[100%] Linking CXX executable openrct2
Undefined symbols for architecture i386:
  "_TTF_CloseFont", referenced from:
      _ttf_dispose in string.c.o
  "_TTF_Init", referenced from:
      _ttf_initialise in string.c.o
  "_TTF_OpenFont", referenced from:
      _ttf_initialise in string.c.o
  "_TTF_Quit", referenced from:
      _ttf_dispose in string.c.o
  "_TTF_RenderUTF8_Solid", referenced from:
      __ttf_surface_cache_get_or_add in string.c.o
  "_TTF_SizeUTF8", referenced from:
      __ttf_getwidth_cache_get_or_add in string.c.o
  "_main", referenced from:
     implicit entry/start for main executable
     (maybe you meant: _window_ride_main_open, _window_main_widgets , _window_main_paint , _window_main_open , _window_get_main , _window_editor_main_open )
  "_platform_directory_delete", referenced from:
      _title_sequence_delete_preset in title_sequences.c.o
  "_platform_ensure_directory_exists", referenced from:
      _cmdline_for_sprite in cmdline_sprite.c.o
      _themes_set_default in config.c.o
      _title_sequences_set_default in config.c.o
      _screenshot_get_next_path in screenshot.c.o
      _title_sequence_create_preset in title_sequences.c.o
      _title_sequence_duplicate_preset in title_sequences.c.o
      _openrct2_initialise in openrct2.c.o
      ...
  "_platform_enumerate_directories_begin", referenced from:
      _title_sequences_load_presets in config.c.o
      _openrct2_copy_files_over in openrct2.c.o
      _window_loadsave_populate_list in loadsave.c.o
  "_platform_enumerate_directories_end", referenced from:
      _title_sequences_load_presets in config.c.o
      _openrct2_copy_files_over in openrct2.c.o
  "_platform_enumerate_directories_next", referenced from:
      _title_sequences_load_presets in config.c.o
      _openrct2_copy_files_over in openrct2.c.o
      _window_loadsave_populate_list in loadsave.c.o
  "_platform_enumerate_files_begin", referenced from:
      _themes_load_presets in config.c.o
      _title_sequence_open in config.c.o
      _object_list_load in object_list.c.o
      _object_list_query_directory in object_list.c.o
      _openrct2_copy_files_over in openrct2.c.o
      _track_load_list in track.c.o
      _track_list_query_directory in track.c.o
      ...
  "_platform_enumerate_files_end", referenced from:
      _themes_load_presets in config.c.o
      _title_sequence_open in config.c.o
      _object_list_load in object_list.c.o
      _object_list_query_directory in object_list.c.o
      _openrct2_copy_files_over in openrct2.c.o
      _track_load_list in track.c.o
      _track_list_query_directory in track.c.o
      ...
  "_platform_enumerate_files_next", referenced from:
      _themes_load_presets in config.c.o
      _title_sequence_open in config.c.o
      _object_list_load in object_list.c.o
      _object_list_query_directory in object_list.c.o
      _openrct2_copy_files_over in openrct2.c.o
      _track_load_list in track.c.o
      _track_list_query_directory in track.c.o
      ...
  "_platform_file_copy", referenced from:
      _game_autosave in game.c.o
      _title_sequence_duplicate_preset in title_sequences.c.o
      _title_sequence_add_save in title_sequences.c.o
      _openrct2_copy_files_over in openrct2.c.o
      _install_track in track.c.o
  "_platform_file_delete", referenced from:
      _theme_delete_preset in themes.c.o
      _title_sequence_create_preset in title_sequences.c.o
      _title_sequence_duplicate_preset in title_sequences.c.o
      _title_sequence_remove_save in title_sequences.c.o
      _track_delete in track.c.o
  "_platform_file_exists", referenced from:
      _cmdline_for_none in cmdline.c.o
      _title_sequence_open in config.c.o
      _game_autosave in game.c.o
      _screenshot_get_next_path in screenshot.c.o
      _title_sequence_add_save in title_sequences.c.o
      _object_load_packed in object.c.o
      _openrct2_copy_files_over in openrct2.c.o
      ...
  "_platform_file_get_modified_time", referenced from:
      _window_loadsave_populate_list in loadsave.c.o
  "_platform_file_move", referenced from:
      _theme_rename_preset in themes.c.o
      _title_sequence_rename_preset in title_sequences.c.o
      _title_sequence_rename_save in title_sequences.c.o
      _track_rename in track.c.o
  "_platform_get_date", referenced from:
      _get_system_time in rct2.c.o
  "_platform_get_drives", referenced from:
      _window_loadsave_populate_list in loadsave.c.o
  "_platform_get_locale_currency", referenced from:
      _config_set_defaults in config.c.o
  "_platform_get_locale_language", referenced from:
      _config_set_defaults in config.c.o
  "_platform_get_locale_measurement_format", referenced from:
      _config_set_defaults in config.c.o
  "_platform_get_locale_temperature_format", referenced from:
      _config_set_defaults in config.c.o
  "_platform_get_time", referenced from:
      _get_local_time in rct2.c.o
  "_platform_get_user_directory", referenced from:
      _config_open_default in config.c.o
      _config_save_default in config.c.o
      _config_shortcut_keys_get_path in config.c.o
      _themes_set_default in config.c.o
      _themes_load_presets in config.c.o
      _themes_save_preset in config.c.o
      _title_sequences_set_default in config.c.o
      ...
  "_platform_open_common_file_dialog", referenced from:
      _show_convert_saved_game_to_scenario_dialog in editor.c.o
      _save_track_design in track.c.o
      _window_loadsave_mouseup in loadsave.c.o
  "_platform_open_directory_browser", referenced from:
      _config_find_or_browse_install_directory in config.c.o
  "_platform_original_game_data_exists", referenced from:
      _config_find_rct2_path in config.c.o
      _rct2_init_directories in rct2.c.o
  "_platform_show_messagebox", referenced from:
      _config_find_or_browse_install_directory in config.c.o
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

janisozaur commented:
It says right there at the bottom:
```
ld: symbol(s) not found for architecture i386
```
Have you installed 32 bit libraries? How does cmake work then, it shouldn't let you compile.

pizza2004 commented:
I mean, there hasn't been a 32 bit kernel /in/ OS X since 2012 at least.  The last OS X to support 32 bit boot was apparently Lion, and even then the last OS X before every machine came with 64 bit installed and used 64 bit was probably Snow Leopard (Snow Leopard was the first fully 64 bit OS X, and since then all machines have shipped with 64 bit kernel).  So like, there hasn't been a version of OS X that was up to date that could even run in 32 bit mode for 5 years or more probably.  There's no reason to target 32 bit.

janisozaur commented:
@pizza2004 I haven't touched OS X in years, but you sound wrong to me. Also your reasoning has major flaws due to the way you confuse kernel and userland.

pizza2004 commented:
I edited my comment a little, but no, I was pretty clear with what I meant.  It hasn't been possible to boot using a 32 bit kernel in OS X on an up to date system since 2011 /at least/.  Perhaps longer than that.  And so all I meant was that, unless we don't have 64 bit versions of a library and we aren't /able/ to build 64 bit versions of a library, it's actually counterintuitive to make it compile in 32 bit when targeting OS X.

I mean, don't get me wrong, if we're working with old code or something and we have to use 32 bit for the time being in order to get it to work then by all means do so, but if we don't have to make it compile in 32 bit on OS X, then there's no need to go through the work of getting it to compile in 32 bit on OS X when later we'll just be switching it over to only a 64 bit version.

janisozaur commented:
OK, I think that sounds much more reasonable now.

In summary your points are: OS X has not had proper 32 bit support since some time ago, as apple is pushing towards 64 bits, so perhaps it would be troublesome (though probably not impossible) to get recent libraries working in 32 bit mode. Basically, a point of practicality and feasibility of solution rather then it being possible at all.

Does that reflect your thoughts?

If that's the case, then I will abstain from judgment, as I am not an OS X user.

I think I could still manage to provide at least some support to those who, if anyone, carries on.

PFCKrutonium commented:
@janisozaur As a hackintosher, I do believe you have the gist of it.

pizza2004 commented:
Yeah, that's pretty much it.  I mean, we could target 32 bit, but given that they haven't built a computer that doesn't have 64 bit support since 2008 (and some before that had it, I know my 2007 MacBook Pro did, anything with an Intel Core 2 Duo or higher does), and every computer since 2010 has been entirely 64 bit through all the hardware, it doesn't make too much sense to target 32 bit for the application.

LinusU commented:
I have built the dependencies as universal binaries so they should work with both 32 and 64 bit. Apples kernel is 64 bit but there is nothing preventing you from running and building 32 bit code.

Is the reason for compiling 32 bit binaries that it won't work with the original binary code otherwise? i.e. that we need to port all functions before we can switch to 64 bit?

Re. my error, that's so strange thought since it files that the compiler just built and they should be 32 bit. They are built with `-m32` and I have also tried `-arch i386` (or something very similar, don't remember the exact command).

LinusU commented:
Tried building it in 64 bit mode and it didn't like the assembly...

```text
In file included from /Users/linus/coding/OpenRCT2/src/game.c:21:
/Users/linus/coding/OpenRCT2/src/audio/../world/../ride/../addresses.h:733:2: error: instruction requires: Not 64-bit mode
        __asm {
        ^
<inline asm>:2:2: note: instantiated into assembly here
        push ebp
        ^
```

pizza2004 commented:
Yeah, I've never encountered issues with building and running in 32 bit, I just meant that there isn't too much point in doing it unless we have to. :D

pizza2004 commented:
I would imagine all the assembly is 32 bit.  So it might be necessary to build it in 32 bit until the program is entirely disassembled, but at the same time, I don't think we'll really get it working properly on OS X until the program is entirely disassembled, so I'm not sure if it is worth the hassle of making it work in 32 bit in the meantime.

LinusU commented:
I definitely think it's worth the hassle, and I don't think that we are too far from it. I would love to have this working as I think it would make it so much easier to develop for OpenRCT2 on Mac. Also I want to play the game and while it actually works under wine with the exe files, it doesn't feel right :smile:

warmwaffles commented:
Checked this branch out on my mac and I get the following errors

```
3 warnings generated.
[ 11%] Building C object CMakeFiles/openrct2.dir/src/hook.c.o
/Users/warmwaffles/code/orct2/src/hook.c:232:91: error: use of undeclared identifier 'MAP_ANONYMOUS'
                g_hooktableaddress = mmap(NULL, size, PROT_EXEC | PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
                                                                                                        ^
1 error generated.
make[2]: *** [CMakeFiles/openrct2.dir/src/hook.c.o] Error 1
make[1]: *** [CMakeFiles/openrct2.dir/all] Error 2
make: *** [all] Error 2
```

janisozaur commented:
I don't have access to osx system, can you point me at some documentation
for mmap? Can you also check that all includes are correct? Especially
`__linux__` vs `_WIN32`, as these can cause a problem in this case.
On Oct 17, 2015 05:33, "Matthew Johnston" <notifications@github.com> wrote:

> Checked this branch out on my mac and I get the following errors
>
> 3 warnings generated.
> [ 11%] Building C object CMakeFiles/openrct2.dir/src/hook.c.o
> /Users/warmwaffles/code/orct2/src/hook.c:232:91: error: use of undeclared identifier 'MAP_ANONYMOUS'
>                 g_hooktableaddress = mmap(NULL, size, PROT_EXEC | PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
>                                                                                                         ^
> 1 error generated.
> make[2]: *** [CMakeFiles/openrct2.dir/src/hook.c.o] Error 1
> make[1]: *** [CMakeFiles/openrct2.dir/all] Error 2
> make: *** [all] Error 2
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRCT2/OpenRCT2/pull/2017#issuecomment-148881935>.
>

LinusU commented:
@warmwaffles Which version of OS X are you on?

warmwaffles commented:
> @warmwaffles Which version of OS X are you on?

latest version, xcode is all up to date.

> Can you also check that all includes are correct? Especially `__linux__` vs `_WIN32`, as these can cause a problem in this case.

I'll poke around tonight and report back.

warmwaffles commented:
Other thing is clang is throwing a ton of warnings stuff like

```
/Users/warmwaffles/code/orct2/src/drawing/string.c:1020:10: warning: initializing 'uint8 *' (aka 'unsigned char *') with an
      expression of type 'char *' converts between pointers to integer types with different sign [-Wpointer-sign]
```

warmwaffles commented:
Upon further inspection, it looks like xcode does not ship with `MAP_ANONYMOUS` defined in the `sys/mman.h`

I did this little hack up top to get some files to compile

```c

#ifdef _WIN32
	#include <windows.h>
#else
	#include <sys/mman.h>
 	#ifndef MAP_ANONYMOUS
 		#define MAP_ANONYMOUS MAP_ANON
 	#endif
#endif // _WIN32
```

warmwaffles commented:
The two files I had to modify

```
	modified:   src/hook.c
	modified:   src/openrct2.c
```

warmwaffles commented:
Speex is also not compiled universally :sob: 

```
ld: warning: ignoring file /usr/local/lib/libspeexdsp.dylib, file was built for x86_64 which is not the architecture being linked (i386): /usr/local/lib/libspeexdsp.dylib
```

pizza2004 commented:
This is why I said it's best not to develop for 32 bit unless it's necessary.  Many people are probably shifting to just 64bit, and many 32bit things won't be bundled anymore. :D

warmwaffles commented:
> This is why I said it's best not to develop for 32 bit unless it's necessary. Many people are probably shifting to just 64bit, and many 32bit things won't be bundled anymore. :D

I agree, but this doesn't help the current situation right now which is the fact we need to support 32 bit until we can get everything decompiled.

pizza2004 commented:
Should I take this to mean that you have an Xcode project set up, or are you just trying to build with the build script as well?  I would be very interested to see the project file being integrated into the code base, as it makes building on Mac much easier.

warmwaffles commented:
I've just been using running `TARGET=darwin ./build.sh`. I have never set up an Xcode project before.

LinusU commented:
I've rebased on master to incorporate all the latest changes, mainly the switch to AT&T syntax, but I'm still hitting the problem with the linking.

@janisozaur I heard that you got a clang build to build, how did you get around this problem: `ld: unknown option: -melf_i386`? It seems like the llvm linker doesn't support that argument. Just removing it from the list (but still keeping `-m32`, should that be enough?) makes the build continue but then it complains it can't find symbols which I think it should.

Some examples of symbols it's complaining on:

```text
Undefined symbols for architecture i386:
  "_TTF_CloseFont", referenced from:
      _ttf_dispose in string.c.o
  "_TTF_Init", referenced from:
      _ttf_initialise in string.c.o
...
  "_main", referenced from:
     implicit entry/start for main executable
     (maybe you meant: _window_ride_main_open, _window_main_widgets , _window_main_paint , _window_main_open , _window_get_main , _window_editor_main_open )
  "_platform_check_steam_overlay_attached", referenced from:
      _platform_create_window in shared.c.o
  "_platform_directory_delete", referenced from:
      _title_sequence_delete_preset in title_sequences.c.o
...
  "_speex_resampler_destroy", referenced from:
      Channel::~Channel() in mixer.cpp.o
  "_speex_resampler_init", referenced from:
      Mixer::MixChannel(Channel&, unsigned char*, int) in mixer.cpp.o
```

The speex thing is probably legit though, it seems like `brew install speex` doesn't support `--universal`...

janisozaur commented:
It does, I made the PR myself: https://github.com/Homebrew/homebrew/pull/45265

your sdl2_ttf looks possibly mis-installed as well.

And you're correct, there's no need to force elf_i386, -m32 is good enough.

Please remove the clang diagnostic push/pop as well, I fixed that.

warmwaffles commented:
heh, @janisozaur I had to hack that in a while back, thanks for making the PR

LinusU commented:
Awesome! I'll just run brew update then :)

Re. sdl2_ttf, the output from lipo looks alright. I'll try a reinstall anyways...

> ➜ ~ lipo -info /usr/local/lib/libSDL2_ttf.dylib
Architectures in the fat file: /usr/local/lib/libSDL2_ttf.dylib are: i386 x86_64 

I think that the strangest part is that it complains on functions that it should have built itself just right before linking them, what do you think about that?

AaronVanGeffen commented:
I'm not convinced we'll be needing the changes with respect to header inclusion. The project builds successfully on [janisozaur's osx branch](https://github.com/janisozaur/OpenRCT2/tree/osx) without these changes. We also run into problems running the resulting executable still, though -- see #1993 for details.

LinusU commented:
Yes that might be right, although I did need to add the `<cerrno>` header to `network.h` to be able to compile his branch.

The one thing our all branches have in common is at least the `<limits.h>` in `draw/rect.h`, I'll make a pull request with only that one. I think a good strategy for this is to slowly inch our way forward to clang compatibility. I would also like to go through all the warnings that clang gives and fix them.

janisozaur commented:
do note my build was configured without network support, so I'd say these headers might be required indeed.

janisozaur commented:
I don't think there's any more code in this branch to be pulled now that #2144 and #2102 are merged.

@LinusU please either rebase & update and I'll make it part of OSX branch or close this.

IntelOrca commented:
I think <limits.h> should only be included once, in rct2.h or common.h. Its a fundamental set of constants that should be available to every source file just like our typedefs.

LinusU commented:
@janisozaur I've rebased and fixed some things (mainly changed more `__linux__` to include OS X as well). This compiles for me, but I still get the linker error that we discussed before.

@IntelOrca There is some discussion of this at the top of this conversation

> **janisozaur**
> personally I'm against globally including headers and would much rather limit them as much as possible. A pass of that in on my mental to-do list. This is partially based on my experience where a project's compilation time was brought down to 10% of original time after cleaning up the headers.

I also agree with that and think that it is good practice to include everything you need in the file you need it in. But if you disagree with this I'll update that as well to be moved to `common.h` (or any other file you want).

janisozaur commented:
why is `__MACH__` required in addition to `__APPLE__`?

LinusU commented:
`__APPLE__` == Mac, `__MACH__` == OS X

[Read more on stackoveflow](http://stackoverflow.com/questions/1529031/what-c-preprocessor-conditional-should-i-use-for-os-x-specific-code)

IntelOrca commented:
@LinusU Less common things like any std container, <list>, <map> or something like <sys/stat.h> should be in the files that uses it.... But <limits.h> will be needed for like over 75% of the source files - and it seems its only OSX that requires it which would decrease performance on other platform compiles. So please put it in the common header.

LinusU commented:
Also, I figured out the linking problems. It's a number of reasons: 1) CMake doesn't pick up SDL2_ttf (see [thie commit for fix](https://github.com/LinusU/OpenRCT2/commit/fb91372940286533d33fbdc52cad48128b05c67d) 2) it can't find libiconv (works when hardcoded in CMakeList.txt) 3) the platform isn't completed for OS X (using the linux one works out of the box, I will make a pull request when this is merged...)

LinusU commented:
@IntelOrca Since `limits.h` has a header guard I don't think it will slow down compile time for other platforms at all?

janisozaur commented:
Can you make this change include https://github.com/janisozaur/OpenRCT2/commit/9fc5f5eb40e6a8aca96b4424e88c885f0b62c5da too?

LinusU commented:
Absolutely, I'll fix it tonight (swedish time)

LinusU commented:
@janisozaur Done!

janisozaur commented:
LGTM. @IntelOrca I'm in favor of this change, will you merge it or do you have other plans with regards to it?

IntelOrca commented:
Only include ```<limits.h>``` in rct2.h - then its good to merge.

LinusU commented:
@IntelOrca Done :+1:

LinusU commented:
Thank you! :fireworks: :shipit:
